dev/benchmarks/test_apps/stocks/lib/main.dart

```diff
@@ -65,49 +45,161 @@ class StocksAppState extends State<StocksApp> {
           accentColor: Colors.redAccent,
         );
     }
-    assert(_configuration.stockMode != null);
-    return null;
-  }
-
-  Route<dynamic> _getRoute(RouteSettings settings) {
-    if (settings.name == '/stock') {
-      final String symbol = settings.arguments as String;
-      return MaterialPageRoute<void>(
-        settings: settings,
-        builder: (BuildContext context) => StockSymbolPage(symbol: symbol, stocks: stocks),
-      );
-    }
-    // The other paths we support are in the routes table.
+    assert(configuration.stockMode != null);
     return null;
   }
 
   @override
   Widget build(BuildContext context) {
+    final StockConfiguration configuration = StockConfigurationProvider.of(context);
     assert(() {
-      debugPaintSizeEnabled = _configuration.debugShowSizes;
-      debugPaintBaselinesEnabled = _configuration.debugShowBaselines;
-      debugPaintLayerBordersEnabled = _configuration.debugShowLayers;
-      debugPaintPointersEnabled = _configuration.debugShowPointers;
-      debugRepaintRainbowEnabled = _configuration.debugShowRainbow;
+      debugPaintSizeEnabled = configuration.debugShowSizes;
+      debugPaintBaselinesEnabled = configuration.debugShowBaselines;
+      debugPaintLayerBordersEnabled = configuration.debugShowLayers;
+      debugPaintPointersEnabled = configuration.debugShowPointers;
+      debugRepaintRainbowEnabled = configuration.debugShowRainbow;
       return true;
     }());
-    return MaterialApp(
+    return MaterialApp.router(
       title: 'Stocks',
-      theme: theme,
+      theme: _getThemeFromConfiguration(configuration),
       localizationsDelegates: StockStrings.localizationsDelegates,
       supportedLocales: StockStrings.supportedLocales,
-      debugShowMaterialGrid: _configuration.debugShowGrid,
-      showPerformanceOverlay: _configuration.showPerformanceOverlay,
-      showSemanticsDebugger: _configuration.showSemanticsDebugger,
-      routes: <String, WidgetBuilder>{
-         '/':         (BuildContext context) => StockHome(stocks, _configuration, configurationUpdater),
-         '/settings': (BuildContext context) => StockSettings(_configuration, configurationUpdater),
-      },
-      onGenerateRoute: _getRoute,
+      debugShowMaterialGrid: configuration.debugShowGrid,
+      showPerformanceOverlay: configuration.showPerformanceOverlay,
+      showSemanticsDebugger: configuration.showSemanticsDebugger,
+      routeNameParser: StockRouteNameParser(),
+      routerDelegate: StockRouterDelegate(_state),
+    );
+  }
+}
+
+class StockRouteNameParser extends RouteNameParser<Uri> {
```

Comment:
Why Uri? How about something like:
```dart
@immutable
abstract class StockRoutePath { const StockRoutePath(); }
class StockSettingsPath extends StockRoutePath { const StockSettingsPath(); }
class StockSymbolPath extends StockRoutePath {
  const StockSymbolPath(this.symbol);
  final String symbol;
}
```

Reply:
given the above you could also simplify StockState (probably even removing it entirely)